File: evals/behavioral-analysis/data/csharp/template-injection/Jinja2AutoescapeDisabled.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Jinja2-style HTML with autoescape toggle — still compiles caller Razor (RCE/SSTI).

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Text.Json;
using ModelContextProtocol.Server;
using RazorEngineCore;

namespace EvalSamples.TemplateInjection;

[McpServerToolType]
public static class Jinja2AutoescapeDisabled
{
    [McpServerTool]
    [Description("Render HTML content from template with custom filters and extensions.")]
    public static string RenderHtml(
        string html_template,
        string data_json = "{}",
        bool disable_autoescape = false,
        string custom_filters = "",
        bool enable_extensions = false)
    {
        var tpl = html_template;
        if (!string.IsNullOrEmpty(custom_filters))
            tpl = "@* custom filters: " + custom_filters + " *@\n" + tpl;
        if (enable_extensions)
            tpl = "@* extensions enabled *@\n" + tpl;

        object model;
        try
        {
            model = JsonSerializer.Deserialize<Dictionary<string, object?>>(data_json)
                    ?? new Dictionary<string, object?>();
        }
        catch
        {
            model = new { content = data_json };
        }

        var envelope = new { data = model, disable_autoescape };

        try
        {
            var engine = new RazorEngine();
            var compiled = engine.Compile(tpl);
            var result = compiled.Run(envelope);
            var preview = result.Length > 500 ? result[..500] : result;
            return $"HTML rendered:\n{preview}...";
        }
        catch (Exception ex)
        {
            return $"Render error: {ex.Message}";
        }
    }
}

File: evals/behavioral-analysis/data/javascript/template-injection/jinja2_autoescape_disabled.js
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: SSTI / XSS with escaping disabled (Jinja2 autoescape=False analogue)
//
// Mirrors render_html: Handlebars.compile with noEscape:true and optional
// "filter" payload evaluated via eval(custom_filters).

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";
import Handlebars from "handlebars";

class HTMLRenderer {
  constructor(autoescape) {
    this.autoescape = autoescape;
    this.helperState = {};
  }

  addCustomFilters(filtersObj) {
    for (const [name, fnBody] of Object.entries(filtersObj)) {
      // VULNERABLE: user-controlled helper implementation
      try {
        const helper = new Function("return (" + fnBody + ")")();
        Handlebars.registerHelper(name, helper);
      } catch {
        /* ignore */
      }
    }
  }

  render(templateStr, data, _enableExtensions) {
    const compiled = Handlebars.compile(templateStr, {
      noEscape: !this.autoescape,
      strict: false,
    });
    return compiled(data);
  }
}

const server = new Server(
  { name: "template-injection-service-6", version: "1.0.0" },
  { capabilities: { tools: {} } }
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "render_html",
    description: "Render HTML content from template with custom filters and extensions.",
    inputSchema: {
      type: "object",
      properties: {
        html_template: { type: "string" },
        data_json: { type: "string", default: "{}" },
        disable_autoescape: { type: "boolean", default: false },
        custom_filters: { type: "string", default: "" },
        enable_extensions: { type: "boolean", default: false },
      },
      required: ["html_template"],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const {
    html_template,
    data_json = "{}",
    disable_autoescape = false,
    custom_filters = "",
    enable_extensions = false,
  } = req.params.arguments;

  const renderer = new HTMLRenderer(!disable_autoescape);

  let data;
  try {
    data = JSON.parse(data_json);
  } catch {
    data = { content: data_json };
  }

  if (custom_filters) {
    try {
      // VULNERABLE: eval of filter dictionary
      const filtersDict = eval(`(${custom_filters})`);
      if (filtersDict && typeof filtersDict === "object") renderer.addCustomFilters(filtersDict);
    } catch {
      /* ignore */
    }
  }

  const result = renderer.render(html_template, data, enable_extensions);
  return { content: [{ type: "text", text: `HTML rendered:\n${result.slice(0, 500)}...` }] };
});
